Support `#[builder(getter(...))]` attribute (#222)
elastio/bon: website/src/reference/builder/member/getter.md

> # `getter` :microscope:

Reviewer: One more thing. There is a table of all attributes in `website/src/reference/builder.md`, it should be updated

> # `overwritable` :microscope:

Reviewer: ```suggestion
# `getter` :microscope:
```

Let's also update the table in refererce/builder.md.

You also need to update the sidebar config in `website/.vitepress/config.mts`. Preserve the lexicographical sorting when adding an entry there.

**Applies to:** <Badge type="warning" text="struct fields"/> <Badge type="warning" text="function arguments"/> <Badge type="warning" text="method arguments"/>

Allows getting a reference to an already set member.

::: danger 🔬 **Experimental**

> This attribute is available under the cargo feature `experimental-getter`. Breaking changes may occur between **minor** releases but not between patch releases.

Reviewer: Don't forget to add this cargo feature into `Cargo.toml` of `bon` and `bon-macros`, and add validation to the code that reads the config that makes sure that feature is enabled before one could use the `builder(getter)` attribute

The fate of this feature depends on your feedback in the tracking issue [#149](https://github.com/elastio/bon/issues/221). Please, let us know if you have a use case for this attribute!

:::

> ::: tip
> 
> This attribute is also configurable via the top-level [`#[builder(on(...))]`](../top-level/on)
> 
> :::

Reviewer: ```suggestion
```

This attribute now makes the following possible:

```rust
#[derive(bon::Builder)]
struct Example {
    #[builder(getter)] // [!code highlight]
    x: u32,
}

let builder = Example::builder().x(1);

let x_ref = builder.get_x(); // [!code highlight]
    
builder.build();
```

This enables builders to hold state that can be accessed by the user while still acting as a builder to some final outcome.
This is especially user if you have a large builder that is getting passed around to many functions,

> and you need to use some previously set value in the builder to set another value.

Reviewer: Let's keep the docs as laconic as possible. `bon`'s documentation is already embarassingly big. I think the use case for getters is already clear, so we can just remove this paragraph. Instead, I'd suggest adding an example of getter for an optional member.

## Config

`getter` is configured quite similarly to [`setters`](./setters).

You can specify any combination of a `name`, `vis`, and `doc` for the getter, or you can omit all of them and use it as a flag to inherit default values for your getters.

### Example

```rust
#[derive(bon::Builder)]
struct Example {
    #[builder(getter(name = get_my_member, vis = "pub(crate)", doc { // [!code highlight]
        /// `get_my_member` gets a ref to the member // [!code highlight]
    }))] // [!code highlight]
    member: u32, // [!code highlight]
}

// Name of `some_fn` that accepts the non-None value was overridden
Example::builder().member(2);

let my_member = builder.get_my_member(); // [!code highlight]
```

## `name`

> The default name for getters are chosen according to the following rules:
> 
> | Member type | Default                                                   |
> | ----------- | --------------------------------------------------------- |
> | Required    | `get_{member}`                                            |
> | Optional    | `get_{member}`                                            |

Reviewer: ```suggestion
The default name for getter is `get_{member}`
```

| Member type | Default                                                   |
| ----------- | --------------------------------------------------------- |
| Required    | `get_{member}`                                            |

> | Optional    | `maybe_get_{member}`                                      |

Reviewer: I don't think there is a need to add a `maybe_` prefix for optional members.
